### PR TITLE
RavenDB-16416: Allow to restore a document when collection was changed

### PR DESCRIPTION
**Pull request by garayx**

### CODE REVIEW

**Comment by grisha-kotler** [comment]
Shouldn't the tombstone that was created fix that? (on import)

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 1]
```diff
@@ -89,32 +89,46 @@ public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase doc
...
-                    if (nonPersistentFlags.Contain(NonPersistentDocumentFlags.FromReplication) == false)
                     {
-                        flags = flags.Strip(DocumentFlags.FromReplication);
-
-                        if (nonPersistentFlags.Contain(NonPersistentDocumentFlags.ByAttachmentUpdate) == false &&
-                            oldFlags.Contain(DocumentFlags.HasAttachments))
-                        {
-                            flags |= DocumentFlags.HasAttachments;
-                        }
+                        if (nonPersistentFlags.Contain(NonPersistentDocumentFlags.FromSmuggler) == false)
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs]: Is there a reason NOT to support it in general?

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs [hunk 2]
```diff
@@ -915,6 +927,12 @@ private SmugglerProgressBase.Counts ProcessTombstones(SmugglerResult result, Bui
                         continue;
                     }
 
+                    if (_isIncrementalRestore && _restoredIds.Contains(tombstone.LowerId))
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs]: Why is this needed?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Exceptions/Documents/CollectionMismatchException.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+using System;
+
+namespace Raven.Client.Exceptions.Documents
+{
+    public class CollectionMismatchException : RavenException
```
Comment (ppekrol) [inline src/Raven.Client/Exceptions/Documents/CollectionMismatchException.cs]: DocumentCollectionMismatchException

File: src/Raven.Client/Exceptions/Documents/CollectionMismatchException.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        public string Id { get; set; }
```
Comment (ppekrol) [inline src/Raven.Client/Exceptions/Documents/CollectionMismatchException.cs]: DocumentId

File: src/Raven.Client/Exceptions/Documents/CollectionMismatchException.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        public CollectionMismatchException()
+        {
+        }
+
+        public CollectionMismatchException(string id, string oldCollection, string newCollection)
+            : base($"Changing '{id}' from '{oldCollection}' to '{newCollection}' via update is not supported. Delete it and recreate the document {id}.")
```
Comment (ppekrol) [inline src/Raven.Client/Exceptions/Documents/CollectionMismatchException.cs]: wrap id in ''

File: src/Raven.Client/Exceptions/Documents/CollectionMismatchException.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+        {
+            Id = id;
+        }
+
+        public CollectionMismatchException(string id, string oldCollection, string newCollection, Exception inner) 
+            : base($"Changing '{id}' from '{oldCollection}' to '{newCollection}' via update is not supported. Delete it and recreate the document {id}.", inner)
```
Comment (ppekrol) [inline src/Raven.Client/Exceptions/Documents/CollectionMismatchException.cs]: wrap id in ''

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 7]
```diff
@@ -556,9 +557,7 @@ private static void ThrowConcurrentExceptionOnMissingDoc(string id, string expec
 
         private static void ThrowInvalidCollectionNameChange(string id, CollectionName oldCollectionName, CollectionName collectionName)
         {
-            throw new InvalidOperationException(
-                $"Changing '{id}' from '{oldCollectionName.Name}' to '{collectionName.Name}' via update is not supported.{Environment.NewLine}" +
-                $"Delete it and recreate the document {id}.");
+            throw new CollectionMismatchException(id, oldCollectionName.Name, collectionName.Name);
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DocumentPutAction.cs]: switch to static factory 'ThrowFor' and put in inside the ex class e.g. like in IndexDoesNotExistException

File: src/Raven.Client/Exceptions/Documents/CollectionMismatchException.cs (hunk 4 above)
Comment (ppekrol) [inline src/Raven.Client/Exceptions/Documents/CollectionMismatchException.cs]: remove set?

File: src/Raven.Client/Exceptions/Documents/CollectionMismatchException.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+using System;
+
+namespace Raven.Client.Exceptions.Documents
+{
+    public class CollectionMismatchException : RavenException
+    {
+        public string Id { get; set; }
+        public CollectionMismatchException()
```
Comment (ppekrol) [inline src/Raven.Client/Exceptions/Documents/CollectionMismatchException.cs]: create set of standard ctors for exceptions and ThrowFor method (check IndexDoesNotExistException)

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 9]
```diff
@@ -163,6 +168,51 @@ public void Dispose()
...
+        public class DuplicateDocsHandler : IDisposable
+        {
+            public List<DocumentItem> DocumentsWithDuplicateCollection;
+            private readonly DocumentsOperationContext _context;
+            private IDisposable _disposable;
+            internal bool _throwOnCollectionMismatchError = false;
+
+            public DuplicateDocsHandler(DocumentDatabase database)
+            {
+                _disposable = database.DocumentsStorage.ContextPool.AllocateOperationContext(out _context);
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: _returnContext

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 10]
```diff
@@ -163,6 +168,51 @@ public void Dispose()
...
+                DocumentsWithDuplicateCollection = new List<DocumentItem>();
+            }
+
+            internal void WriteDocumentWithDuplicateCollection(DocumentItem item)
+            {
+                var doc = new Document
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: create a method inside document `Clone(JsonOperationContext context)` and handle everything there. Less error prone in future.

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 11]
```diff
@@ -163,6 +168,51 @@ public void Dispose()
...
+            private IDisposable _disposable;
+            internal bool _throwOnCollectionMismatchError = false;
+
+            public DuplicateDocsHandler(DocumentDatabase database)
+            {
+                _disposable = database.DocumentsStorage.ContextPool.AllocateOperationContext(out _context);
+                DocumentsWithDuplicateCollection = new List<DocumentItem>();
+            }
+
+            internal void WriteDocumentWithDuplicateCollection(DocumentItem item)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: rename to `AddDocument`

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 12]
```diff
@@ -844,12 +915,14 @@ public void Dispose()
...
         {
             public bool IsRevision;
+            public bool ThrowOnCollectionMismatchError;
 
             private readonly DocumentDatabase _database;
             private readonly BuildVersionType _buildType;
             private readonly Logger _log;
 
             public readonly List<DocumentItem> Documents = new List<DocumentItem>();
+            public List<DocumentItem> DocumentsWithDuplicateCollection = new List<DocumentItem>();
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: create that list only if necessary, this is a very rare case.

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs [hunk 13]
```diff
@@ -250,6 +251,22 @@ private void ProcessType(DatabaseItemType type, SmugglerResult result, BuildVers
             _onProgress.Invoke(result.Progress);
         }
 
+        private void TryProcessDocumentsWithDuplicateCollection(SmugglerResult result, BuildVersionType buildType)
+        {
+            if (_destination is DatabaseDestination databaseDestination && databaseDestination._duplicateDocsHandler.DocumentsWithDuplicateCollection.Count > 0)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs]: If you need to do the type check here then it is modelled incorrectly.

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs [hunk 14]
```diff
@@ -578,13 +595,13 @@ private SmugglerProgressBase.Counts ProcessRevisionDocuments(SmugglerResult resu
             return result.RevisionDocuments;
         }
 
-        private SmugglerProgressBase.Counts ProcessDocuments(SmugglerResult result, BuildVersionType buildType)
+        private SmugglerProgressBase.Counts ProcessDocuments(SmugglerResult result, BuildVersionType buildType, IEnumerable<DocumentItem> docs = null)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs]: Move to ProcessDocumentInternal and in ProcessDocuments use previous behavior and use ProcessDocumentInternal in the second place.

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs [hunk 15]
```diff
@@ -250,6 +251,22 @@ private void ProcessType(DatabaseItemType type, SmugglerResult result, BuildVers
...
 
+        private void TryProcessDocumentsWithDuplicateCollection(SmugglerResult result, BuildVersionType buildType)
+        {
+            if (_destination is DatabaseDestination databaseDestination && databaseDestination._duplicateDocsHandler.DocumentsWithDuplicateCollection.Count > 0)
+            {
+                using (databaseDestination._duplicateDocsHandler)
+                {
+                    result.AddInfo(
+                        $"Starting to process '{databaseDestination._duplicateDocsHandler.DocumentsWithDuplicateCollection.Count}' documents with duplicate collection.");
+                    var localResult = new SmugglerResult();
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs]: Why this is like that? Can you explain?

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs [hunk 16]
```diff
@@ -250,6 +251,22 @@ private void ProcessType(DatabaseItemType type, SmugglerResult result, BuildVers
             _onProgress.Invoke(result.Progress);
         }
 
+        private void TryProcessDocumentsWithDuplicateCollection(SmugglerResult result, BuildVersionType buildType)
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs]: Remove Try

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs (hunk 13 above)
Comment (ppekrol) [reply src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs]: introduce in destination interface method: `GetDocumentsWithDuplicateCollection` and introduce a parameter to actions.Documents for handling the duplicates instead of changing some internal flag.

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 17]
```diff
@@ -257,6 +320,9 @@ public void Dispose()
                 FinishBatchOfDocuments();
                 FixDocumentMetadataIfNecessary();
                 DeleteRevisionsForNonExistingDocuments();
+
+                if (_duplicateDocsHandler._markForDispose)
```
Comment (arekpalinski) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:326]: Why are we disposing if _only_ if it's marked?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 18]
```diff
@@ -1040,8 +1129,18 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
...
+                        _database.DocumentsStorage.Put(context, id, expectedChangeVector: null, document.Data, document.LastModified.Ticks, document.ChangeVector, document.Flags, document.NonPersistentFlags);
+                    }
+                    catch (DocumentCollectionMismatchException)
+                    {
+                        if (ThrowOnCollectionMismatchError)
+                            throw;
 
-                    _database.DocumentsStorage.Put(context, id, null, document.Data, document.LastModified.Ticks, document.ChangeVector, document.Flags, document.NonPersistentFlags);
+                        DocumentsWithDuplicateCollection ??= new List<DocumentItem>();
+                        DocumentsWithDuplicateCollection.Add(documentType);
```
Comment (arekpalinski) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: Maybe instead of adding two fields - `ThrowOnCollectionMismatchError` and `DocumentsWithDuplicateCollection` just add a handler like:

`Action<DocumentItem> DocumentCollectionMismatchHandler` (default: null)

and then throw if the handler is null, otherwise call `DocumentCollectionMismatchHandler.Invoke(doc)`.

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 19]
```diff
@@ -445,12 +515,29 @@ private void FinishBatchOfDocuments()
...
                 }
 
                 _command = null;
             }
+
+            private void BatchPutWithDuplicateCollectionHandling(MergedBatchPutCommand command, Task task)
+            {
+                AsyncHelpers.RunSync(() => task);
+
+                if (command.DocumentsWithDuplicateCollection == null)
```
Comment (arekpalinski) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: This is a bit strange IMO that _after_ executing the command we're checking this `DocumentsWithDuplicateCollection` collection and make additional actions on it. See my comment below where I proposed to have dedicated `DocumentCollectionMismatchHandler`

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs (hunk 17 above)
Comment (garayx) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:326]: this is to dispose it right after the tombstones type is processed, generally it always get disposed (not only if marked) see line 67.

Comment (arekpalinski) [reply src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:326]: ok

**Review by arekpalinski — APPROVED** [review]

**Review by ppekrol — APPROVED** [review]
